# Title: Cleaning Dataset using Python library - "Pandas"
## Author: Daniyal Siddiqui

### Purpose of the Project:
- To use the pandas library to efficiently cleanup a large messy dataset to be used for data analysis later.
- To find out the most expensive cars in the dataset.
---

In this analysis we will use the dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The original dataset can be found [here](https://www.kaggle.com/orgesleka/used-cars-database/data). We will be working with a modified version of it. The modifications were done as follows:
- 50,000 data points from the full dataset were sampled
- The dataset was dirtied a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

---

We will star off by reading the file into the Pandas dataframe:

In [249]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "Latin-1") # Reading the file in default encoding raised an error.

We will now explore the dataset:

In [250]:
print(autos.info())
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can make the following observations from the dataset:
- Multiple columns have digits and characters mixed in
- A couple of column contain date and time
- Multiple columns have null values.
- Column names are in [Camel Case](https://en.wikipedia.org/wiki/Camel_case) notation rather than the [Snake case](https://en.wikipedia.org/wiki/Snake_case) form.

## Cleaning up column names:
Lets deal with the column names first:

In [251]:
autos.rename(columns = {'yearOfRegistration':'registration_year','monthOfRegistration':'registration_month',
              'notRepairedDamage':'unrepaired_damage','dateCreated':'ad_created','dateCrawled':'date_crawled',
                       'offerType':'offer_type','vehicleType':'vehicle_type','powerPS':'power_PS',
                       'fuelType':'fuel_type','nrOfPictures':'nr_of_pictures','postalCode':'postal_code',
                       'lastSeen':'last_seen'},inplace = True)
autos.head(1)

date_crawled                              name  seller offer_type  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat    Angebot   

    price   abtest vehicle_type  registration_year  gearbox  power_PS   model  \
0  $5,000  control          bus               2004  manuell       158  andere   

    odometer  registration_month fuel_type    brand unrepaired_damage  \
0  150,000km                   3       lpg  peugeot              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54

## Exploring descriptive statistics of all columns:

In [252]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From the above descriptive statistics of all columns we can make the following observations:
- The columns with names `seller` and `offer_type` has mostly a single repeated value (i.e., `freq` = 49999), therefore they can be dropped from the further aalysis, as they will not produce any meaninigful value.
- The column with names `price` and `odometer` should be numeric but they looks to be stored as text.

## Coverting mixed values columns to numeric only:
Lets convert the `price` and `odometer` column values to `float` first by removing the special and string characters.

In [253]:
autos['price'] = autos['price'].str.replace('\W','').astype(float)
autos['odometer'] = autos['odometer'].str.replace('km|\W','').astype(float)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_PS   model  odometer  registration_month fuel_type  \
0    manuell       158  andere  150000.0                   3       lpg   
1  automatik       286     7er  150000.0                   6    benzin   
2    manuell       102    golf   70000.0                   7    benzin   
3  automatik        71  fortwo   70000.0                   6    benzin   
4    manuell         0   focus  150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Lets analyze the price and odometer column a bit more by exploring the entries in each of the column. We will again use the describe() method since now they are converted to floats.

In [254]:
autos.rename(columns = {'price':'price_$','odometer':'odometer_km'}, inplace = True)
print("Statistics for the price column are:")
print(autos['price_$'].describe(),'\n')
print("Statistics for the odometer column are:")
print(autos['odometer_km'].describe())

Statistics for the price column are:
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64 

Statistics for the odometer column are:
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


## Taking out the outliers:
We can see that the maximum and minimum values of te price column donot makes sense, so we will explore it further to remove the outliers. we will set the threshold of valid car prices to a minimum of $1000 and a max of $100,000:

In [255]:
autos2 = autos.loc[autos['price_$'].between(1000,100000), ]
print("Statistics for the new price column are:")
print(autos2['price_$'].describe(),"\n")
print("The length of the rows removed from the dataset are: ",len(autos)-len(autos2))

Statistics for the new price column are:
count    38587.000000
mean      7089.045482
std       7893.814349
min       1000.000000
25%       2200.000000
50%       4300.000000
75%       8950.000000
max      99900.000000
Name: price_$, dtype: float64 

The length of the rows removed from the dataset are:  11413


## Exploring the date-time columns:
We will now move on to the five columns representing the date and time. These are as follows:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

Firstly, we will calculate the distribution of values in the `date_crawled`, `ad_created`, and `last_seen` columns (all string columns). Secondly we will print the statistics for the `registration_month` and `registration_year`column:

In [256]:
string_dates = ['date_crawled','last_seen','ad_created']
for i in string_dates:
    print("The first entry is",autos2[i].value_counts(dropna = True).sort_index(ascending = True).head(1))
    print("The last entry is",autos2[i].value_counts(dropna = True).sort_index(ascending = True).tail(1),'\n')
numeric_dates = ['registration_month','registration_year']
for i in numeric_dates:
    print('The earliest entry in {0} is'.format(i), autos2[i].min())
    print('The latest entry in {0} is'.format(i), autos2[i].max(),'\n')

The first entry is 2016-03-05 14:06:30    1
Name: date_crawled, dtype: int64
The last entry is 2016-04-07 14:36:56    1
Name: date_crawled, dtype: int64 

The first entry is 2016-03-05 14:46:02    1
Name: last_seen, dtype: int64
The last entry is 2016-04-07 14:58:50    3
Name: last_seen, dtype: int64 

The first entry is 2015-06-11 00:00:00    1
Name: ad_created, dtype: int64
The last entry is 2016-04-07 00:00:00    51
Name: ad_created, dtype: int64 

The earliest entry in registration_month is 0
The latest entry in registration_month is 12 

The earliest entry in registration_year is 1000
The latest entry in registration_year is 9999 



One thing that stands out from the exploration we did in the last cell is that the registration_year column contains some odd values:

* The minimum value is 1000, before cars were invented
* The maximum value is 9999, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [257]:
autos3 = autos2.loc[autos2["registration_year"].between(1900,2016), ]
autos3["registration_year"].describe()

count    37167.000000
mean      2003.967552
std          7.008246
min       1927.000000
25%       2001.000000
50%       2005.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

The dataset makes much more sense now with the oldest car model of `1927` and the latest of `2016`. Also the dataset is further shortened to `37167` entries now.

**The dataset is now cleanly prepared to do some statistical or exploratory analysis**

## Finding out the most expensive cars:
We will now shortened our dataset to the 10 most recurring car brands and do some analysis on them:

In [258]:
most_recurring_brands = autos3.loc[:,'brand'].value_counts().head(10)
print("The most recurring car brands in order are: ",'\n',most_recurring_brands)
most_recurring_brands = most_recurring_brands.index

The most recurring car brands in order are:  
 volkswagen       7843
bmw              4661
mercedes_benz    4149
audi             3630
opel             3314
ford             2183
renault          1387
peugeot          1038
fiat              784
skoda             709
Name: brand, dtype: int64


Lets find out the mean prices of the newly found most recurring car brands:

In [259]:
mean_price = {}
for i in most_recurring_brands:
    mean_price[i] = autos3.loc[autos3['brand']==i, 'price_$'].mean()
print("The mean price of the most recurring vehicles in the dataset are: ")
print(sorted(mean_price.items(), key=lambda x: x[1], reverse=True))

The mean price of the most recurring vehicles in the dataset are: 
[('audi', 10276.903581267217), ('mercedes_benz', 9196.060737527116), ('bmw', 8975.103840377602), ('skoda', 6836.696755994359), ('volkswagen', 6645.13260232054), ('ford', 5274.369216674301), ('opel', 4219.954737477368), ('fiat', 4008.174744897959), ('peugeot', 3955.169556840077), ('renault', 3590.942321557318)]


We will now check to see if the change in prices between the top 10 brands has anything to do with the mileage they have:

In [260]:
milleage = {}
for i in most_recurring_brands:
    milleage[i] = autos3.loc[autos3['brand']==i, 'odometer_km'].mean()
print("The mean odometer reading (milleage) of the most recurring vehicles in the dataset are: ")
print(sorted(milleage.items(), key=lambda x: x[1], reverse=True))

The mean odometer reading (milleage) of the most recurring vehicles in the dataset are: 
[('bmw', 132068.2257026389), ('mercedes_benz', 130130.15184381779), ('audi', 127524.79338842975), ('volkswagen', 125765.65089889073), ('opel', 123952.92697646348), ('peugeot', 122341.04046242774), ('renault', 121423.93655371305), ('ford', 119651.85524507558), ('skoda', 110063.46967559944), ('fiat', 107901.78571428571)]


Lets join the two dictionaries in a dataframe to compare better:

In [261]:
mean_price_series = pd.Series(mean_price)
milleage_series = pd.Series(milleage)
price_milleage_df = pd.DataFrame(mean_price_series, columns=['mean_price'])
price_milleage_df['mean_milleage'] = milleage_series
price_milleage_df['price/mile'] = price_milleage_df['mean_price']/price_milleage_df['mean_milleage']
price_milleage_df.sort_values('price/mile', ascending = False)

mean_price  mean_milleage  price/mile
audi           10276.903581  127524.793388    0.080587
mercedes_benz   9196.060738  130130.151844    0.070668
bmw             8975.103840  132068.225703    0.067958
skoda           6836.696756  110063.469676    0.062116
volkswagen      6645.132602  125765.650899    0.052837
ford            5274.369217  119651.855245    0.044081
fiat            4008.174745  107901.785714    0.037147
opel            4219.954737  123952.926976    0.034045
peugeot         3955.169557  122341.040462    0.032329
renault         3590.942322  121423.936554    0.029574

## Results:
- We can see that the brand named `audi` is indeed the most expensive car among the five most recurring cars in the dataframe and its value is not effected by the odometer rating.
- Brand named `opel` has mean price more than the brand named `fiat`, but this is because of the much lower milleage `fiat` have ran.